### PR TITLE
devufs can now read the superblock when mounting

### PR DESCRIPTION
**Pull request by gmacd**
Signed-off-by: Graham MacDonald <grahamamacdonald@gmail.com>

### CODE REVIEW

File: sys/include/libufs.h [hunk 1]
```diff
@@ -10,11 +10,22 @@
...
 struct Chan;
-struct mount;
 struct vnode;
+struct thread;
 
+/* Wrapper for a UFS mount.  Should support reading from both kernel and user
+ * space (eventually)
+ */
+struct mount {
+	struct ufsmount*	mnt_data;
```
Comment (rminnich) [inline sys/include/libufs.h]: we need to start making this more like Plan 9 code.
typedef Struct ufsmount *Ufsmount
then declare it as Ufsmount *mnt_data;

File: sys/include/libufs.h [hunk 2]
```diff
@@ -10,11 +10,22 @@
...
+	struct Chan*		chan;
```
Comment (rminnich) [inline sys/include/libufs.h]: Chan *chan;

File: sys/include/libufs.h [hunk 3]
```diff
@@ -10,11 +10,22 @@
...
+	int32_t			(*read)(struct mount*, void*, int32_t, int64_t);
```
Comment (rminnich) [inline sys/include/libufs.h]: what's the reason for this? You have the chan. Why is that not sufficient?

File: sys/include/libufs.h [hunk 4]
```diff
@@ -10,11 +10,22 @@
...
+};
+
+struct mount* newufsmount(
+	struct Chan* c,
```
Comment (rminnich) [inline sys/include/libufs.h]: Chan *c, not struct.

File: sys/include/libufs.h [hunk 5]
```diff
@@ -10,11 +10,22 @@
...
+	int32_t (*read)(struct mount*, void*, int32_t, int64_t));
 
-struct mount* newufsmount(struct Chan* c);
 struct vnode* newufsvnode();
```
Comment (rminnich) [inline sys/include/libufs.h]: You need to create
typedef struct vnode Vnode.

Or better struct Vnode.
typedef struct Vnode Vnode;

File: sys/include/libufs.h [hunk 6]
```diff
@@ -10,11 +10,22 @@
...
 
 void releaseufsmount(struct mount* mp);
```
Comment (rminnich) [inline sys/include/libufs.h]: Mount *mp

File: sys/src/9/port/devufs.c [hunk 7]
```diff
@@ -138,10 +138,16 @@ ufsread(Chan *c, void *a, int32_t n, int64_t offset)
 	return n;
 }
 
+static int32_t
+chanread(struct mount* mp, void *p, int32_t n, int64_t off)
```
Comment (rminnich) [inline sys/src/9/port/devufs.c]: I don't get why you need this. Why isn't dev->read enough? 

FYI the Plan 9 way to do this is you store the type in the QID and case out on that in dev->read.

File: sys/src/9/port/devufs.c [hunk 8]
```diff
@@ -138,10 +138,16 @@ ufsread(Chan *c, void *a, int32_t n, int64_t offset)
...
+{
+	return mp->chan->dev->read(mp->chan, p, n, off);
+}
+
 static struct mount*
 mountufs(Chan* c)
 {
-	struct mount* mp = newufsmount(c);
+	struct mount* mp = newufsmount(c, chanread);
```
Comment (rminnich) [inline sys/src/9/port/devufs.c]: Mount * but anyway we need to revisit all this chanread stuff.

File: sys/src/libufs/ffs/ffs_vfsops.c [hunk 9]
```diff
@@ -38,17 +38,19 @@
...
 //#include "../ffs/ffs_extern.h"
 
 #include "libufs.h"
+#include "ufs_harvey.h"
 
-#if 0
-static uma_zone_t uma_inode, uma_ufs1, uma_ufs2;
+//static uma_zone_t uma_inode, uma_ufs1, uma_ufs2;
 
 static int	ffs_mountfs(struct vnode *, struct mount *, struct thread *);
```
Comment (rminnich) [inline sys/src/libufs/ffs/ffs_vfsops.c]: Vnode *, Mount *, Thread *

This is not linux :-)

File: sys/src/libufs/ffs/ffs_vfsops.c [hunk 10]
```diff
@@ -85,9 +87,10 @@ static const char *ffs_opts[] = { "acls", "async", "noatime", "noclusterr",
 int
 ffs_mount(struct mount *mp)
 {
+	struct vnode *devvp = nil;
```
Comment (rminnich) [inline sys/src/libufs/ffs/ffs_vfsops.c]: Moutn *
Vnode *

well you get my drift by now I bet.

File: sys/src/libufs/ufs_harvey.c [hunk 11]
```diff
@@ -17,11 +17,14 @@
 
 
 struct mount*
-newufsmount(Chan* c)
+newufsmount(
+	struct Chan* c,
```
Comment (rminnich) [inline sys/src/libufs/ufs_harvey.c]: Chan *c

File: sys/include/libufs.h (hunk 3 above)
Comment (gmacd) [inline sys/include/libufs.h]: I'm trying to make the library usable for kernel or user code.  I could only see a way to get access to the channel's internal dev member (in portdat.h) in the kernel code.  Also, I assumed people would usually open a file outside the kernel using 'open'.  This way permits either, assuming we eventually make the Chan a union with a fid (or do something else).

File: sys/include/libufs.h (hunk 1 above)
Comment (gmacd) [inline sys/include/libufs.h]: Ok - I'll look at a spatch to rename ufsmount in a later PR and make the typedef change for this one.

File: sys/include/libufs.h (hunk 5 above)
Comment (gmacd) [inline sys/include/libufs.h]: I'll make a spatch PR for this soon.

File: sys/src/libufs/ffs/ffs_vfsops.c (hunk 10 above)
Comment (gmacd) [inline sys/src/libufs/ffs/ffs_vfsops.c]: Hmm, one more time?

File: sys/include/libufs.h (hunk 3 above)
Comment (rminnich) [inline sys/include/libufs.h]: Were it me, I would make it work for one thing, and worry about the other later when we've figured out how it all works. The reason is that this is getting kinda nasty at this point, doing things you don't want to do in kernel for the eventual and maybe never happens user mode case. My belief is we should make this right for kernel first, and then figure out the user case if and when we need it.

Comment (gmacd) [inline sys/include/libufs.h]: Alright - let me look into it.

**Comment by elbing** [comment]
looks fine for me now.

File: sys/include/libufs.h [hunk 12]
```diff
@@ -9,15 +9,29 @@
  */
 
 
-struct Chan;
-struct mount;
-struct vnode;
+typedef struct Chan Chan;
```
Comment (rminnich) [inline sys/include/libufs.h]: it's weird that you need this but ...

**Comment by gmacd** [comment]
I haven't finished yet - still working on removing the read field.

**Pull request closed without merging**